quipucords/api/credential/serializer.py

```diff
@@ -23,225 +22,180 @@ def expand_filepath(filepath):
-    )
-    become_method = ValidStringChoiceField(
-        required=False, choices=Credential.BECOME_METHOD_CHOICES
-    )
-    become_user = CharField(required=False, max_length=64)
-    become_password = CharField(
-        required=False,
-        max_length=1024,
-        allow_null=True,
-        style={"input_type": "password"},
-    )
+    """Base Serializer for the Credential model."""
 
     class Meta:
         """Metadata for the serializer."""
 
         model = Credential
         fields = "__all__"
+        extra_kwargs = {
+            "password": ENCRYPTED_FIELD_KWARGS,
+            "auth_token": ENCRYPTED_FIELD_KWARGS,
+            "ssh_passphrase": ENCRYPTED_FIELD_KWARGS,
+            "become_password": ENCRYPTED_FIELD_KWARGS,
+        }
 
-    def validate(self, attrs):
-        """Validate if fields received are appropriate for each credential."""
-        cred_type = get_from_object_or_dict(self.instance, attrs, "cred_type")
-
-        if cred_type == DataSources.VCENTER:
-            validated_data = self.validate_vcenter_cred(attrs)
-        elif cred_type == DataSources.SATELLITE:
-            validated_data = self.validate_satellite_cred(attrs)
-        elif cred_type == DataSources.NETWORK:
-            validated_data = self.validate_host_cred(attrs)
-        elif cred_type == DataSources.OPENSHIFT:
-            validated_data = self.validate_openshift_cred(attrs)
-        else:
-            raise ValidationError({"cred_type": messages.UNKNOWN_CRED_TYPE})
-        return validated_data
-
-    def create(self, validated_data):
-        """Create host credential."""
-        name = validated_data.get("name")
-        check_for_existing_name(
-            Credential.objects, name, _(messages.HC_NAME_ALREADY_EXISTS % name)
+    # Instead of following DRF BaseSerializer.__new__ signature (cls, *args, **kwargs),
+    # our __new__ method uses BaseSerializer.__init__ signature. This is OK, because:
+    # A) class level __new__ just calls __init__ with the same *args, *kwargs [1];
+    # B) DRF BaseSerializer.__new__ only modifies kwargs [3];
+    #
+    # References
+    # [1]: https://docs.python.org/3/reference/datamodel.html#object.__new__
+    # [2]: https://github.com/encode/django-rest-framework/blob/3.14.0/rest_framework/serializers.py#L121  # noqa: E501
+
+    def __new__(
+        cls, instance=None, data=empty, **kwargs
+    ):  # pylint: disable=arguments-differ
+        """Overloaded __new__ to return the appropriate serializer."""
+        cred_type = cls._get_cred_type(instance, data)
+        klass = cls._get_serializer_class(cred_type)
+        return super().__new__(klass, instance=instance, data=data, **kwargs)
+
+    @classmethod
+    def _get_serializer_class(cls, cred_type):
+        """Return the appropriate serializer based on 'cred_type'."""
+
+        def _default():
+            return CredentialSerializer
+
+        serializer_per_datasource = defaultdict(
+            _default,
```

Comment:
Minor question: Was there a reason other than preferred coding style for writing this as a function instead of an inline anonymous lambda expression like this?

```suggestion
        serializer_per_datasource = defaultdict(
            lambda: CredentialSerializer,
```

Reply:
just PTSD (I thought I was clever writing lots of lambdas for processing a HUGE amount of data - the machine I had access had 1Tb ram XD - until I realized lambdas were causing that implementation to be slow as hell and had to refactor)

for web development this makes zero difference and is even more readable. I'll addopt your suggestion :wink: